프로젝트 매니저 [STEP 2-1] Moon

hojun-jo opened the pull request:
@havilog
안녕하세요 하비!
프로젝트 매니저 STEP 2-1 PR 입니다!
조언 해주신 것을 참고하여 테이블 뷰 하나에서 데이터를 보여주는 것을 MVVM으로 구현해 보았습니다.


## 고민했던 점
### Cell에서 데이터 바인딩
Cell은 재사용 되고 어떤 액션도 없는 상태이기 때문에 데이터를 받아오는 방법을 고민했습니다.
재사용 될 때마다 ViewModel을 주입받아 `listCellViewModel` 프로퍼티가 바뀔 때마다 바인딩 할 수 있도록 구현했습니다.
```swift
// ListCell이 재사용 될 때마다 setUpViewModel을 통해 값이 들어오면 바인딩 함
private var listCellViewModel: ListCellViewModel? {
    didSet {
        setUpBindings()
    }
}
// ListCell이 재사용 될 때 ListCellViewModel을 받아옴
func setUpViewModel(_ viewModel: ListCellViewModel) {
    self.listCellViewModel = viewModel
}
// viewModel을 받을 때마다 바인딩할 것
private func setUpBindings() {
    listCellViewModel?.bindTitle { [weak self] viewModel in
        self?.titleLabel.text = viewModel.title
    }
    
    listCellViewModel?.bindDescription { [weak self] viewModel in
        self?.descriptionLabel.text = viewModel.description
    }
    
    listCellViewModel?.bindDeadline { [weak self] viewModel in
        self?.deadlineLabel.text = viewModel.deadline
    }
}
```

### 헤더와 데이터 바인딩
테이블 뷰의 헤더에 목록의 개수를 표시하는 부분이 있어 뷰 모델과 바인딩이 필요하다고 생각했습니다.
따라서 다음과 같이 모델이 변경 되었을 때 헤더가 바인딩하고 있는 프로퍼티도 바뀔 수 있도록 구현했습니다.
```swift
final class ListViewModel {
    // ListViewController에서 ListCell의 ListCellViewModel로 전달하게될 모델의 배열
    private var todoList: [Todo] {
        didSet {
            count = String(todoList.count)
        }
    }
    // ListHeader의 contentAmountLabel와 바인딩할 변수
    var count: String {
        didSet {
            bindCount?(self)
        }
    }
    // ListHeader의 contentAmountLabel와 바인딩하기 위한 클로저
    private var bindCount: ((ListViewModel) -> Void)?
...
}
```


## 조언을 얻고 싶은 점
### 모델 생성 시점
현재는 `ListViewModel`을 초기화하며 임의로 데이터를 넣어둔 상태입니다.
```swift
final class ListViewModel {
...
    init() {
        todoList = [
            Todo(title: "1", description: "111", deadline: Date()),
            Todo(title: "2", description: "222", deadline: Date()),
            Todo(title: "3", description: "333", deadline: Date()),
            ...
        ]
        
        count = String(todoList.count)
    }
...
}
```
하지만 저장소에 데이터를 저장 해두고 모델에서 불러와 사용한다고 했을 때 어느 시점에 초기화하는 것이 좋을 지 잘 모르겠습니다.
제가 생각한 방법은 다음과 같습니다.
1. `SceneDelegate`에서 모델 생성 -> 이 모델을 주입 받아 뷰 모델 생성 -> 이 뷰 모델을 주입 받아 뷰 컨트롤러 생성
- 이 경우 `SceneDelegate`에서 데이터를 가져오는 로직을 수행하게 되는 것 같아 적절하지 않은 것 같습니다.
2. 뷰 모델에서 모델을 초기화하며 저장소에서 데이터를 가져오는 로직 수행
- 뷰 모델은 프레젠테이션 로직을 수행하는 것으로 알고 있는데 데이터를 가져오는 로직을 수행하게 되는 것은 적절하지 않은 것 같습니다.
3. 모델 자체에서 초기화되며 데이터를 가져오는 로직 수행
- 데이터를 가지고 있고 비즈니스 로직을 수행 해야하는 모델이 직접 이 작업을 하는 것이 괜찮은 것인지 모르겠습니다.


### 뷰 모델의 역할
테이블 뷰의 `numberOfRowsInSection`가 `todoList`의 개수가 되어야하고 `cellForRowAt`에서는 `Todo` 모델이 필요하기 때문에 뷰 모델에서 이를 리턴하고 있습니다.
뷰 모델의 역할은 프레젠테이션 로직을 수행하는 것인데 이런 로직을 수행해도 괜찮은 부분일까요?
```swift
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return listViewModel.countTodo()
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: ListCell.identifier) as? ListCell else {
            return UITableViewCell()
        }
        
        let todo = listViewModel.getTodo(index: indexPath.row)
        
        cell.setUpViewModel(ListCellViewModel(todo: todo))
        
        return cell
    }
```

Commit:
chore: MVVM 폴더 구조 생성

Commit:
refactor: ViewController에서 ListViewController로 네이밍 수정

Commit:
feat: ListViewController에 tableView 추가

Commit:
chore: ListViewController 파일 이름 수정

Commit:
feat: ListViewCell 생성 및 구현

Commit:
feat: ListViewHeader 생성 및 구현

Commit:
refactor: ListViewCell의 subView들을 contentView의 subView로 수정

Commit:
feat: ListViewController의 TableView에 Cell 사이 간격 추가

Commit:
refactor: 뷰들의 배경색 수정

Commit:
refactor: ListViewCell의 deadlineLabel 네이밍 수정

Commit:
feat: Todo 모델 생성

Commit:
feat: ListViewModel 생성

Commit:
refactor: ListHeader, ListCell 네이밍 수정

Commit:
feat: identifier를 위한 Reusable 프로토콜 생성

Commit:
feat: ListCellViewModel 생성

Commit:
refactor: ListHeader와 바인딩할 수 있도록 수정

- todoList가 변경되면 ListHeader와 바인딩할 count가 바뀔 수 있도록 구현
- ListHeader와 바인딩할 count 변수 생성

Commit:
feat: ListCell과 ListCellViewModel 바인딩

Commit:
feat: ListHeader와 ListViewModel 바인딩

Commit:
refactor: ListViewController에서 ViewModel을 사용할 수 있도록 수정

havilog commented on ProjectManager/ProjectManager/Model/Todo.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+struct Todo {
+    var title: String
+    var description: String
+    var deadline: Date
```
요 아이들을 variable로 선언하신 이유가 있을까요?

hojun-jo replied:
할 일을 수정할 수 있기 때문에 계속 변경될 거라 생각해 variable로 선언했습니다!

havilog replied:
사실 프로퍼티의 값이 변경될 일이 잦으면 class가 더 적합한 거 같고,
할 일을 수정하는건 할 일을 담당하는 객체가 가지고 있는 todo의 인스턴스가 변경될거 같아요

hojun-jo replied:
아아 변경될 일이 잦으면 class가 더 적합하겠군요
코어데이터를 사용하면 지금의 구조체를 사용하지 않고 NSManagedObject인 todo를 사용해서 값이 변경되면 뷰모델이 데이터매니저에게 저장하도록 만들 것 같습니다!

havilog commented on ProjectManager/ProjectManager/View/ListCell.swift:
```diff
@@ -0,0 +1,127 @@
+        return stackView
+    }()
+    
+    // ListCell이 재사용 될 때마다 setUpViewModel을 통해 값이 들어오면 바인딩 함
+    private var listCellViewModel: ListCellViewModel? {
+        didSet {
+            setUpBindings()
+        }
+    }
```
사실 셀에 뷰모델이 필요한가? 는 조금 논쟁거리였는데요,

제 기준을 설명드리자면

셀 자체적으로 API를 쏘고 캐싱을 하는 등의 side effect가 날 수 있는 로직이 들어가면, 
외부(뷰컨)에서 뷰모델을 만들어서 주입해주는 편이구요,

그저 데이터를 보여주기만 하는 셀이라면 모델(Todo)를 주입받아
그 모델의 데이터만 뿌려주는 편이에요

hojun-jo replied:
https://github.com/yagom-academy/ios-project-manager/pull/307/commits/0752eea359905763d58f58eb4e06ac2dd8a85782

셀에서 데이터를 보여주기만 하기 때문에 뷰모델을 사용하지 않도록 하는 것이 효율적인 것 같아 수정했습니다!

havilog commented on ProjectManager/ProjectManager/View/ListViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        return ListHeader(viewModel: listViewModel)
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return listViewModel.countTodo()
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: ListCell.identifier) as? ListCell else {
+            return UITableViewCell()
+        }
+        
+        let todo = listViewModel.getTodo(index: indexPath.row)
+        
+        cell.setUpViewModel(ListCellViewModel(todo: todo))
+        
+        return cell
+    }
```
사실 뷰 모델의 todos 배열 자체에 접근 할 수 있어도 괜찮을거 같아요.
사실 count나 index에 있는 아이템을 바로 접근하는 것 보다는 현재 방식이
정석에 더 가까운 방식이긴 한데요
얻는 이점에 비해 오버엔지니어링으로 손해가 조금 더 클 것 같아서요

hojun-jo replied:
https://github.com/yagom-academy/ios-project-manager/pull/307/commits/0752eea359905763d58f58eb4e06ac2dd8a85782
좀 더 간단하게 사용할 수 있게 배열 자체에 접근할 수 있도록 수정했습니다!

havilog commented on ProjectManager/ProjectManager/ViewModel/ListViewModel.swift:
```diff
@@ -0,0 +1,53 @@
+    private var bindCount: ((ListViewModel) -> Void)?
+    
+    init() {
+        todoList = [
+            Todo(title: "1", description: "111", deadline: Date()),
+            Todo(title: "2", description: "222", deadline: Date()),
+            Todo(title: "3", description: "333", deadline: Date()),
+            Todo(title: "4", description: "444", deadline: Date()),
+            Todo(title: "5", description: "555", deadline: Date()),
+            Todo(title: "6", description: "666", deadline: Date())
+        ]
```
고민하신 부분처럼 init에서 넣어주기 보다는,

뷰컨에서는 viewDidLoad 시점을 쏴주고,
뷰모델에서 뷰디드로드 이벤트를 받아,
데이터 스토어에서 데이터를 로드하는걸 시도하고,
메모리에 값을 넣거나 에러 핸들링하는 로직이 들어가면 될 것 같습니다.

hojun-jo replied:
https://github.com/yagom-academy/ios-project-manager/pull/307/commits/588d15484560a53ce51db47c7da001b67fe1a07e
아직 코어 데이터 활용 전이라 데이터 로드 시점만 수정했습니다!
코어 데이터를 사용하게 되면 이 시점에 데이터 로드를 시도할 수 있도록 수정하겠습니다 😀

havilog reviewed (approved):
수고하셨어요

Commit:
feat: TodoDateFormatter, DateFormat 생성

Commit:
refactor: Cell에서 ViewModel을 사용하지 않도록 수정

Commit:
refactor: todoList 모델 로드 시점 수정

ListViewController에서는 viewDidLoad 시점을 쏴주고,
ListViewModel에서 viewDidLoad 이벤트를 받아,
데이터 로드를 시도할 수 있도록 수정

Commit:
feat: Array extension 생성 및 safe subscript 구현

Commit:
refactor: 배열 접근을 safe subscript로 하도록 수정
